(slv-124) ability to redact sensitive env var values in rototiller

er0ck opened the pull request.

Commit:
(SLV-124) add ability to redact sensitive env var values in rototiller

* add a new EnvVar type for sensitive values
  * remove `#default` functionality so we don't promote saving sensitive values in repos
  * redact values in logging
* add spec tests on the above
* add `#add_env_sensitive` to rototiller_task
* add `#add_env_sensitive` to RototillerParam
* remove add_hash_param as it was specific to non sensitive envs

the sensitive type is added to the existing env_collection array as this
simplifies messaging, stopping, etc.  and its the whole value in using a
subtype.

Commit:
(maint) abstract #add_env into intermediate class

prior to this change #add_env and its supporting functions were
scattered through the other params like `Command` and `Switch`.  This
made adding a sensitive env type difficult as we would have to further
propagate tests for the new env to all those params.
This change creates an intermediate type on top of `RototillerParam`
which handles `#add_env`

Commit:
(maint) move duplicated `#validate_hash` method to hash_handling

Commit:
(SLV-124) bump to a Y version because it's a feature!

samwoods1 commented:
I have a general question.. Do we print out the cmd by default with all of the values? I don't see anywhere we do that, but I know we talked about it.
Right now, to_str or to_s would show the value of the sensitive env_var, which probably wouldn't be desired, BUT we're using to_str to actually construct and execute the command and it has to be the actual value there.  Do we need a different, private method to construct the actual command vs to_str which is public and should redact the sensitive values?

er0ck commented:
> I have a general question.. Do we print out the cmd by default with all of the values? I don't see anywhere we do that, but I know we talked about it.
> Right now, to_str or to_s would show the value of the sensitive env_var, which probably wouldn't be desired, BUT we're using to_str to actually construct and execute the command and it has to be the actual value there. Do we need a different, private method to construct the actual command vs to_str which is public and should redact the sensitive values?

good question.  maybe. there are lots of ways rake task authors can get around this and print the value including `puts ENV['blah']` so i'm not sure how much i want to protect them from themselves, including restricting `puts env_sensitive`.
also, a `rototiller_task` is not a `RototillerParameter` so it wouldn't have access to a private method (and probably not a protected one).  although i guess a task should never have to print it (nor be able to). which i guess is your point?  hmmm....

also, i don't think i adequately test that a command can still get a sensitive env var's value...  (i'll fix that up).  oh also, the command is printed with all its values, that is probably still broken if you use a sensitive env var to override a command or parameter.

er0ck commented:
crap.  this makes printing the command before its run kinda difficult.  right now we don't know whether an env_var overrode a parameter name at all, let alone if that override is sensitive.  to complicate this further, we could have any number of env_vars attached to an override.  i can plumb this through, but it'll be more work.

we only just recently added printing the command.  i view it as pretty useful, but i could just remove that feature, i suppose (that'd be breaking)

billclaytor commented on docs/env_var_example_reference.md:
```diff
@@ -42,8 +41,30 @@ Because a task has a name that would not be implicitly useful as the default val
       end
     end
 
+### A task author wants input from the user, but not to print the environment's value
```
This might be more clear as something like: "A task author wants input from the user, but does not want to print the environment variable's value"

Commit:
(SLV-124) redact printed command strings

In the previous commits we redact in the param messaging/logging.
But by default, and as an option, you can spit out the command that is
to be run, for easier debugging and transparency.  This may contain
sensitive env var values (which replaced param values).
So we needed a way to print the command string in a safe way, while
still retaining those values to be run silently in the background.  so
here we split out #to_s/to_str with #safe_print.

We had to add an object attribute to track whether a param's value came
from a sensitive env var.

Commit:
(maint) add new env_var stubs to allow env_var mocks and pry in spec tests

pry continuously adds new env vars for its interface. If we don't stub
them out and pass them through, our mocks for env vars break pry. this
allows debugging with pry as of now.

Commit:
(SLV-124) add spec tests for safe_print and is_value_sensitive

* test safe_print functionality and ensure supporting attributes behave
correctly.

er0ck commented:
> > I have a general question.. Do we print out the cmd by default with all of the values? I don't see anywhere we do that, but I know we talked about it.
> > Right now, to_str or to_s would show the value of the sensitive env_var, which probably wouldn't be desired, BUT we're using to_str to actually construct and execute the command and it has to be the actual value there. Do we need a different, private method to construct the actual command vs to_str which is public and should redact the sensitive values?
> 
> good question. maybe. there are lots of ways rake task authors can get around this and print the value including `puts ENV['blah']` so i'm not sure how much i want to protect them from themselves, including restricting `puts env_sensitive`.
> also, a `rototiller_task` is not a `RototillerParameter` so it wouldn't have access to a private method (and probably not a protected one). although i guess a task should never have to print it (nor be able to). which i guess is your point? hmmm....
> 
> also, i don't think i adequately test that a command can still get a sensitive env var's value... (i'll fix that up). oh also, the command is printed with all its values, that is probably still broken if you use a sensitive env var to override a command or parameter.

ok. i got this portion done and tested.  works well and i think the solution is _decent_

er0ck commented:
oops. forgot to check rubocop, etc

Commit:
(maint) ensure we don't copy the Gemfile.lock into CI container

* this used to be a problem, then it wasn't. now bundler2 incompats
with possible 1.x bundler inside the container make it a problem again.
* if you run with CI=true locally, this will delete your lockfile

er0ck commented on lib/rototiller/task/params/command.rb:
```diff
@@ -165,19 +130,34 @@ def add_argument(*args, &block)
+          my_value = "[REDACTED]"
+        elsif name
+          my_value = name
+        end
```
weird that rubocop didn't catch this.  this should probably be assigned based on the output of the conditional, instead of two possible assignments inside the conditional. 🤷‍♂️
